src/extensions/extensionsHelper.ts

```diff
@@ -534,43 +650,156 @@ async function validateExtensionSpec(args: {
         "Please add one so users know what has changed in this version. " +
         marked(
           "See https://firebase.google.com/docs/extensions/alpha/create-user-docs#writing-changelog for more details."
         )
     );
   }
+  return notes;
+}
 
-  if (args.latestVersion) {
-    if (semver.lt(extensionSpec.version, args.latestVersion)) {
+/**
+ * Validates the extension version.
+ *
+ * @param extensionRef the ref of the extension
+ * @param newVersion the new extension version
+ * @param latestVersion the latest extension version
+ */
+function validateVersion(extensionRef: string, newVersion: string, latestVersion?: string) {
+  if (latestVersion) {
+    if (semver.lt(newVersion, latestVersion)) {
       throw new FirebaseError(
         `The version you are trying to publish (${clc.bold(
-          extensionSpec.version
+          newVersion
         )}) is lower than the current version (${clc.bold(
-          args.latestVersion
+          latestVersion
         )}) for the extension '${clc.bold(
           extensionRef
-        )}'. Please make sure this version is greater than the current version (${clc.bold(
-          args.latestVersion
-        )}) inside of extension.yaml.\n`,
+        )}'. Make sure this version is greater than the current version (${clc.bold(
+          latestVersion
+        )}) inside of extension.yaml and try again.\n`,
         { exit: 104 }
       );
-    } else if (semver.eq(extensionSpec.version, args.latestVersion)) {
+    } else if (semver.eq(newVersion, latestVersion)) {
       throw new FirebaseError(
-        `The version you are trying to publish (${clc.bold(
-          extensionSpec.version
-        )}) already exists for Extension '${clc.bold(
+        `The version you are trying to upload (${clc.bold(
+          newVersion
+        )}) already exists for extension '${clc.bold(
           extensionRef
-        )}'. Please increment the version inside of extension.yaml.\n`,
+        )}'. Increment the version inside of extension.yaml and try again.\n`,
         { exit: 103 }
       );
     }
   }
+}
 
-  return { extensionSpec, notes };
+/** Unpacks extension state into a more specific string. */
+export function unpackExtensionState(extension: Extension) {
+  switch (extension.state) {
+    case "PUBLISHED":
+      // Unpacking legacy "published" terminology.
+      if (extension.latestApprovedVersion) {
+        return clc.bold(clc.green("Published"));
+      } else if (extension.latestVersion) {
+        return clc.green("Uploaded");
+      } else {
+        return "Prerelease";
+      }
+    case "DEPRECATED":
+      return clc.red("Deprecated");
+    case "SUSPENDED":
+      return clc.bold(clc.red("Suspended"));
+    default:
+      return "-";
+  }
+}
+
+/**
+ * Fetches the extension and its latest version and displays all metadata.
+ *
+ * @param extensionRef the ref of the extension
+ */
+async function displayExtensionHeader(extensionRef: string): Promise<{
```

Comment:
Would be better to not combine the logic of displaying a header with the logic for fetching backend code. 

I'd prefer to make this only handle the logging, and do the fetching elsewhere:
```
async function displayExtensionHeader(e: Extension, latestEv?: ExtensionVersion): Promise<void>
```